kore/src/Kore/Exec.hs

```diff
@@ -603,6 +615,56 @@ matchDisjunction mainModule matchPattern disjunctionPattern =
     sort = Pattern.patternSort matchPattern
     match = Search.matchWith SideCondition.top
 
+{- | Ensure that for every equation in a function definition,
+ the right-hand side of the equation is a function pattern.
+ 'checkFunctions' first extracts equations from a verified module
+ to a list of equations. Then it 'traverse's the list, checking
+ that each equation is a function pattern. If any equation fails
+ the check, 'checkFunctions' logs an error message containing
+ the offending equation. After checking each equation, if any
+ have failed the check then 'checkFunctions' returns @'ExitFailure' 3@;
+ otherwise it returns 'ExitSuccess'.
+
+ See 'checkEquation',
+ 'Kore.Equation.Registry.extractEquations',
+ and 'Kore.Internal.TermLike.isFunctionPattern'.
+-}
+checkFunctions ::
+    MonadLog m =>
+    -- | The main module
+    VerifiedModule StepperAttributes ->
+    m ExitCode
+checkFunctions verifiedModule = checkResults <$> traverse checkEquation equations
+  where
+    equations = join $ Map.elems $ extractEquations verifiedModule
+    -- if any equations fail the check,
+    -- the entire function returns ExitFailure 3.
+    checkResults results
+        | ExitFailure 3 `elem` results = ExitFailure 3
+        | otherwise = ExitSuccess
+
+{- | 'checkEquation' returns 'ExitSuccess' when the 'right'-hand-side
+ of an 'Equation' is a function pattern. Otherwise, log the error message
+ "RHS of equation is not a function pattern:", pretty-print the
+ offending equation, and return @'ExitFailure' 3@.
+
+ See 'Kore.Internal.TermLike.isFunctionPattern'.
+-}
+checkEquation ::
+    MonadLog m =>
+    InternalVariable variable =>
+    Equation variable ->
+    m ExitCode
+checkEquation eqn@Equation{right}
+    | isFunctionPattern right = return ExitSuccess
+    | otherwise = do
+        logError $ renderText $ layoutOneLine err
```

Comment:
Let's add a new log type for this (see [Kore.Log](https://github.com/kframework/kore/tree/master/kore/src/Kore/Log) for examples).